lib/fa2/nft/nft.impl.jsligo

```diff
@@ -0,0 +1,232 @@
+   //  operators 
+   /**
+* Check if the intented transfer is sent from the same sender as from field, otherwise check if the sender is part of the operator authorized to receive this token
+* @param operators :  operator bigmap
+* @param from_ : transfer from address
+* @param token_id : token_id to test  
+*/
+
+   export const assert_authorisation = (
+      operators: operators,
+      from_: address,
+      token_id: nat
+   ): unit => {
+      const sender_ = (Tezos.get_sender());
+      if (sender_ != from_) {
+         const authorized =
+            match((Big_map.find_opt([from_, sender_], operators))) {
+               when (Some(a)):
+                  a
+               when (None()):
+                  Set.empty
+            };
+         if (! (Set.mem(token_id, authorized))) {
+            return failwith(Errors.not_operator)
+         }
+      } else {
+         return unit
+      }
+   };
+   export const add_operator = (
+      operators: operators,
+      owner: address,
+      operator: operator,
+      token_id: nat
+   ): operators => {
+      if (owner == operator) {
+         return operators
+      // assert_authorisation always allow the owner so this case is not relevant
+
+      } else {
+         Assertions.assert_update_permission(owner);
+         let auth_tokens =
+            match(Big_map.find_opt([owner, operator], operators)) {
+               when (Some(ts)):
+                  ts
+               when (None()):
+                  Set.empty
+            };
+         auth_tokens = Set.add(token_id, auth_tokens);
+         return Big_map.update([owner, operator], Some(auth_tokens), operators)
+      }
+   };
+   export const remove_operator = (
+      operators: operators,
+      owner: address,
+      operator: operator,
+      token_id: nat
+   ): operators => {
+      if (owner == operator) {
+         return operators
+      // assert_authorisation always allow the owner so this case is not relevant
+
+      } else {
+         Assertions.assert_update_permission(owner);
+         const auth_tokens: option<set<nat>> =
+            match(Big_map.find_opt([owner, operator], operators)) {
+               when (Some(ts)):
+                  do {
+                     const toks = Set.remove(token_id, ts);
+                     if (Set.cardinal(toks) == 0n) {
+                        return None()
+                     } else {
+                        return Some(toks)
+                     }
+                  }
+               when (None()):
+                  None()
+            };
+         return Big_map.update([owner, operator], auth_tokens, operators)
+      }
+   }
+   //  ledger 
+
+   export const is_owner_of = (ledger: ledger, token_id: nat, owner: address): bool => {
+      const current_owner = Option.unopt(Big_map.find_opt(token_id, ledger));
+      return (current_owner == owner)
+   };
+   const assert_owner_of = (ledger: ledger, token_id: nat, owner: address): unit =>
```

Comment:
I think this one should be in assertions

Reply:
you cannot generalize this one as ledger definition is different on each implementation